File: logback-access-spring-boot-starter/src/main/java/net/rakugakibox/spring/boot/logback/access/undertow/LogbackAccessUndertowHttpHandler.java
package net.rakugakibox.spring.boot.logback.access.undertow;

import io.undertow.server.ExchangeCompletionListener.NextListener;
import io.undertow.server.HttpHandler;
import io.undertow.server.HttpServerExchange;
import io.undertow.server.handlers.accesslog.AccessLogHandler;
import net.rakugakibox.spring.boot.logback.access.LogbackAccessContext;
import net.rakugakibox.spring.boot.logback.access.LogbackAccessProperties;
import org.springframework.context.ApplicationEventPublisher;

/**
 * The Undertow HTTP handler that emits Logback-access events.
 * This class is own implementation from scratch, based on {@link AccessLogHandler}.
 * Because emits an own customized access event ({@link UndertowLogbackAccessEvent}).
 *
 * @see AccessLogHandler
 */
public class LogbackAccessUndertowHttpHandler implements HttpHandler {

    /**
     * The Logback-access context.
     */
    private final LogbackAccessContext logbackAccessContext;

    /**
     * The next HTTP handler.
     */
    private final HttpHandler nextHandler;

    /**
     * Constructs an instance.
     *
     * @param logbackAccessProperties the configuration properties for Logback-access.
     * @param applicationEventPublisher the application event publisher.
     * @param nextHandler the next HTTP handler.
     */
    public LogbackAccessUndertowHttpHandler(
            LogbackAccessProperties logbackAccessProperties,
            ApplicationEventPublisher applicationEventPublisher,
            HttpHandler nextHandler
    ) {
        this.logbackAccessContext = new LogbackAccessContext(logbackAccessProperties, applicationEventPublisher);
        this.nextHandler = nextHandler;
        this.logbackAccessContext.configure();
        this.logbackAccessContext.start();
    }

    /** {@inheritDoc} */
    @Override
    public void handleRequest(HttpServerExchange exchange) throws Exception {
        exchange.addExchangeCompleteListener(this::handleExchangeEvent);
        nextHandler.handleRequest(exchange);
    }

    /** {@inheritDoc} */
    public void handleExchangeEvent(HttpServerExchange exchange, NextListener nextListener) {
        UndertowLogbackAccessEvent event = new UndertowLogbackAccessEvent(exchange);
        logbackAccessContext.emit(event);
        nextListener.proceed();
    }

}

File: src/main/kotlin/dev/akkinoc/spring/boot/logback/access/undertow/LogbackAccessUndertowHttpHandler.kt
package dev.akkinoc.spring.boot.logback.access.undertow

import dev.akkinoc.spring.boot.logback.access.LogbackAccessContext
import dev.akkinoc.spring.boot.logback.access.LogbackAccessEvent
import io.undertow.server.ExchangeCompletionListener.NextListener
import io.undertow.server.HttpHandler
import io.undertow.server.HttpServerExchange
import org.slf4j.Logger
import org.slf4j.LoggerFactory.getLogger

/**
 * The Undertow [HttpHandler] to emit Logback-access events.
 *
 * @property logbackAccessContext The Logback-access context.
 * @property next The next [HttpHandler].
 * @see io.undertow.server.handlers.accesslog.AccessLogHandler
 */
class LogbackAccessUndertowHttpHandler(
        private val logbackAccessContext: LogbackAccessContext,
        private val next: HttpHandler,
) : HttpHandler {

    override fun handleRequest(exchange: HttpServerExchange) {
        log.debug(
                "Handling the {}: {} @{}",
                HttpServerExchange::class.simpleName,
                exchange,
                logbackAccessContext,
        )
        exchange.addExchangeCompleteListener(::log)
        next.handleRequest(exchange)
    }

    /**
     * Logs the request/response exchange.
     *
     * @param exchange The request/response exchange.
     * @param next The next listener.
     */
    private fun log(exchange: HttpServerExchange, next: NextListener) {
        log.debug(
                "Logging the {}: {} @{}",
                HttpServerExchange::class.simpleName,
                exchange,
                logbackAccessContext,
        )
        val source = LogbackAccessUndertowEventSource(
                exchange = exchange,
                localPortStrategy = logbackAccessContext.properties.localPortStrategy,
        )
        val event = LogbackAccessEvent(source)
        logbackAccessContext.emit(event)
        next.proceed()
    }

    companion object {

        /**
         * The logger.
         */
        private val log: Logger = getLogger(LogbackAccessUndertowHttpHandler::class.java)

    }

}
